Source/Dafny/Compilers/SinglePassCompiler.cs

```diff
@@ -548,20 +548,66 @@ protected interface IClassWriter {
     /// <summary>
     /// Emit a loop like this:
     ///     foreach (tmpVarName:collectionElementType in [[collectionWriter]]) {
-    ///       if (tmpVarName is member of type boundVarType) {
-    ///         var boundVarName:boundVarType := tmpVarName as boundVarType;
-    ///         [[bodyWriter]]
-    ///       }
+    ///       [[bodyWriter]]
     ///     }
     /// where
     ///   * "[[collectionWriter]]" is the writer returned as "collectionWriter"
     ///   * "[[bodyWriter]]" is the block writer returned
+    /// </summary>
+    protected abstract ConcreteSyntaxTree CreateForeachLoop(
+      string tmpVarName, Type collectionElementType,
+      Bpl.IToken tok, out ConcreteSyntaxTree collectionWriter, ConcreteSyntaxTree wr);
+
+    /// <summary>
+    /// Creates a guarded foreach loop that iterates over a collection, and apply required subtype
+    /// and compiled subset types filters. Will not emit intermediate ifs if there is no need.
+    ///
+    ///     foreach(collectionElementType tmpVarName in collectionWriter) {
+    ///       if(tmpVarName is [boundVar.type]) {
+    ///         var [IDName(boundVar)] = ([boundVar.type])(tmpvarName);
+    ///         if(constraints_of_boundvar.Type([IDName(boundVar)])) {
+    ///           ...
+    ///         }
+    ///       }
+    ///     }
+    /// </summary>
+    /// <returns>A writer to write inside the deepest if-then</returns>
+    private ConcreteSyntaxTree CreateGuardedForeachLoop(
+      string tmpVarName, Type collectionElementType,
+      IVariable boundVar,
+      bool introduceBoundVar, bool inLetExprBody,
+      Bpl.IToken tok, out ConcreteSyntaxTree collectionWriter, ConcreteSyntaxTree wr
+      ) {
+      wr = CreateForeachLoop(tmpVarName, collectionElementType, tok, out collectionWriter, wr);
+      wr = MaybeInjectSubtypeConstraint(tmpVarName, collectionElementType, boundVar.Type, inLetExprBody, tok, wr);
+      EmitDowncastVariableAssignment(IdName(boundVar), boundVar.Type, tmpVarName, collectionElementType,
+          introduceBoundVar, tok, wr);
+      wr = MaybeInjectSubsetConstraint(boundVar, boundVar.Type, collectionElementType, inLetExprBody, tok, wr);
+      return wr;
+    }
+
+    /// <summary>
+    /// Returns a subtype condition like:
+    ///     tmpVarName is member of type boundVarType
+    /// Returns null if no condition is necessary
+    /// </summary>
+    [CanBeNull]
+    protected abstract string GetSubtypeCondition(
+      string tmpVarName, Type boundVarType, Bpl.IToken tok, ConcreteSyntaxTree wPreconditions);
+
+    /// <summary>
+    /// Emit an (already verified) downcast assignment like:
+    ///     var boundVarName:boundVarType := tmpVarName as boundVarType;
+    ///     [[bodyWriter]]
+    /// </summary>
+    /// where
+    ///   * "[[bodyWriter]]" is the block writer returned
     /// Option:
```

Comment:
```cs
    /// <summary>
    /// Emit an (already verified) downcast assignment like:
    /// 
    ///     var boundVarName:boundVarType := tmpVarName as boundVarType;
    ///     [[bodyWriter]]
    /// 
    /// where
    ///   * "[[bodyWriter]]" is where the writer wr's position will be next
    /// </summary>
    /// <param name="boundVarName">Name of the variable after casting</param>
    /// <param name="boundVarType">Expected variable type</param>
    /// <param name="tmpVarName">The collection's variable name</param>
    /// <param name="collectionElementType">type this variable is casted from, in case it is useful</param>
    /// <param name="introduceBoundVar">Whether or not to declare the variable, in languages requiring declarations</param>
    /// <param name="tok">A position in the AST</param>
    /// <param name="wr">The concrete syntax tree writer</param>
    ```